File: Mehul-Singhal/PrefixSUM/Problem3.java
import java.util.Scanner;
class Problem3 {
        int[][] matrix;
        public Problem3(int[][] matrix) {
              int rows = matrix.length;// no of rows
            int colm = matrix[0].length;//no of columns in first roe
    
            //O(1) TIME COMPLEXITY
    
            for (int i = 1; i < rows; i++) {
                matrix[i][0] += matrix[i - 1][0];
            }
            for (int i = 1; i < colm; i++) {
                matrix[0][i] += matrix[0][i - 1];
            }
    
              for(int i = 1; i<rows ; i++){
                for(int j =1; j<colm ;j++){ 
                    matrix[i][j] += matrix[i][j - 1] + matrix[i - 1][j] - matrix[i - 1][j - 1];
                }
              }
                this.matrix =matrix;
        }
            
        public int sumRegion(int row1, int col1, int row2, int col2) {

            int top =0,left=0,topleft=0;
            if(row1>0){
                top = matrix[row1-1][col2];
            }
            if(col1>0){
                left = matrix[row2][col1-1];
            }
            if(row1>0 && col1>0){
                topleft = matrix[row1-1][col1-1];
            }
            return matrix[row2][col2] - top -left +topleft;
        } 
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter input array length: ");
        int inputLen = sc.nextInt();

        String inputArray[] = new String[inputLen];

        for(int i=0; i<inputLen; i++){
            inputArray[i] = sc.next();
        }

        System.out.print("Enter row of matrix: ");
        int m= sc.nextInt();


        System.out.print("Enter col of matrix: ");
        int n = sc.nextInt();

        int matrix[][] = new int[m][n];
        System.out.print("Enter elements of matix: ");
        
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                matrix[i][j] = sc.nextInt();
            }
        }

        int ans[] = new int[inputLen-1];

        Problem3 numMatrix = new Problem3(matrix);

        for(int i=1; i<inputArray.length; i++){
            System.out.print("Enter range for sum: ");
            System.out.print("Enter range for sum: ");
            int row1 = sc.nextInt();
            int row2 = sc.nextInt();
            int col1 = sc.nextInt();
            int col2 = sc.nextInt();
            ans[i-1] = numMatrix.sumRegion(row1, col1, row2, col2);
        }
        sc.close();
        for(int i:ans){
            System.out.print(i + " ");
        }
    }
}

File: Aditya Pratap SIngh/PrefixSumAssessment/Problem3.py
class NumMatrix:
    def __init__(self, matrix):
        if not matrix:
            return

        m=len(matrix)
        n=len(matrix[0])
        self.prefix= [[0] * (n + 1) for _ in range(m + 1)]#empty matrix of same size

        for i in range(m):
            for j in range(n):
                self.prefix[i + 1][j + 1] = (matrix[i][j] + self.prefix[i][j + 1] +
                                             self.prefix[i + 1][j] - self.prefix[i][j])

    def sumRegion(self, row1, col1, row2, col2):
        return (self.prefix[row2 + 1][col2 + 1] - self.prefix[row1][col2 + 1] - self.prefix[row2 + 1][col1] + self.prefix[row1][col1])

# Input operations
operations= ["NumMatrix", "sumRegion", "sumRegion", "sumRegion"]
inputs= [[[[3, 0, 1, 4, 2],
            [5, 6, 3, 2, 1],
            [1, 2, 0, 1, 5],
            [4, 1, 0, 1, 7],
            [1, 0, 3, 0, 5]]], 
          [2, 1, 4, 3], 
          [1, 1, 2, 2], 
          [1, 2, 2, 4]]


methods= {"NumMatrix": lambda inp: NumMatrix(*inp),
    "sumRegion": lambda inp: num_matrix.sumRegion(*inp)}

result=[]
num_matrix = None

for op, inp in zip(operations, inputs):
    if op=="NumMatrix":
        num_matrix= methods[op](inp)
        result.append(None)
    else:
        result.append(methods[op](inp))

print(result)
